### PR TITLE
SNOW-1641729: Support diff() for timestamp columns.

### PR DESCRIPTION
**Pull request by sfc-gh-mvashishtha**
Fixes SNOW-1641729

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -64,6 +64,7 @@
   - support for `pd.to_timedelta`.
   - support for `GroupBy` aggregations `min`, `max`, `mean`, `idxmax`, `idxmin`, `std`, `sum`, `median`, `count`, `any`, `all`, `size`, `nunique`.
   - support for `TimedeltaIndex` attributes: `days`, `seconds`, `microseconds` and `nanoseconds`.
+  - support for `diff()` with timestamp columns on `axis=0` and `axis=1`
```
Comment (sfc-gh-azhan) [inline CHANGELOG.md:67]: can we update docs too?

Comment (sfc-gh-azhan) [reply CHANGELOG.md:67]: nvm

Comment (sfc-gh-azhan) [inline CHANGELOG.md:67]:
```suggestion
  - support for `diff` with timestamp columns on `axis=0` and `axis=1`
```

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 2]
```diff
@@ -10558,8 +10579,12 @@ def diff(self, periods: int, axis: int) -> "SnowflakeQueryCompiler":
         }
         return SnowflakeQueryCompiler(
             self._modin_frame.update_snowflake_quoted_identifiers_with_expressions(
-                diff_label_to_value_map,
-                self._modin_frame.cached_data_column_snowpark_pandas_types,
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10562]: In which case, that the result of diff has different types than the original df?

Comment (sfc-gh-mvashishtha) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10562]: diff() between two timestamp columns on axis=1, or on a timestamp column on axis 0, will change type to timedelta. I added a comment.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 3]
```diff
@@ -10491,15 +10492,25 @@ def _make_discrete_difference_expression(
                         self._modin_frame.ordering_column_snowflake_quoted_identifiers
                     )
                 )
-                return (col1 | col2) & (not_(col1 & col2))
+                return SnowparkPandasColumn(
+                    snowpark_column=(col1 | col2) & (not_(col1 & col2)),
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10496]: This is interesting! I tracked it here https://snowflakecomputing.atlassian.net/browse/SNOW-1650680

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 4]
```diff
@@ -10485,15 +10486,25 @@ def _make_discrete_difference_expression(
...
+                    snowpark_pandas_type=None,
+                )
             else:
-                return col(snowflake_quoted_identifier) - func_for_other(
-                    snowflake_quoted_identifier, offset=abs(periods)
-                ).over(
-                    Window.order_by(
-                        self._modin_frame.ordering_column_snowflake_quoted_identifiers
-                    )
+                return compute_binary_op_between_snowpark_columns(
```
[the remark this answers is not in the document]
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10500]: This is a great start! We should extend this to other places calling binary ops.

**Review by sfc-gh-azhan — APPROVED** [review]

**Pull request merged**